File: editcompetitorsdialog.cpp
#include "editcompetitorsdialog.h"
#include "ui_editcompetitorsdialog.h"

EditCompetitorsDialog::EditCompetitorsDialog(QList<ParticipantInfo>& participants, QWidget *parent) :
    QDialog(parent),
    ui(new Ui::EditCompetitorsDialog),
    m_participants(participants)
{
    ui->setupUi(this);

    updateCompetitorsTable();
}

EditCompetitorsDialog::~EditCompetitorsDialog()
{
    delete ui;
}

void EditCompetitorsDialog::updateCompetitorsTable ()
{
    // Set up the model of the QTableView
    QStandardItemModel *model = new QStandardItemModel(m_participants.length()+1,3,this); //2 Rows and 3 Columns

    model->setHorizontalHeaderItem(0, new QStandardItem(QString("Name")));
    model->setHorizontalHeaderItem(1, new QStandardItem(QString("Car Name")));
    model->setHorizontalHeaderItem(2, new QStandardItem(QString("Delete")));

    for (int i = 0 ; i < m_participants.length() ; i++)
    {
        ParticipantInfo part = m_participants.at(i);

        model->setData(model->index(i, 0), part.participantName());
        model->setData(model->index(i, 1), part.carName());
    }
    ui->participantsTable->setModel(model);

    for (int i = 0 ; i < m_participants.length() ; i++)
    {
        ParticipantInfo part = m_participants.at(i);

        QPushButton* delButton = new QPushButton();
        delButton->setText("Delete");

        ui->participantsTable->setIndexWidget(model->index(i , 2), delButton);

        connect(delButton, &QPushButton::released, [=]{
            handleDeleteButton(i);
        });
    }

    connect(ui->participantsTable->model(), SIGNAL(dataChanged(const QModelIndex&, const QModelIndex&)), this, SLOT(onDataChanged(const QModelIndex&, const QModelIndex&)));

}

void EditCompetitorsDialog::onDataChanged(const QModelIndex& topLeft, const QModelIndex& bottomRight)
{
    int row = topLeft.row();
    int col = topLeft.column();
    if (row == ui->participantsTable->model()->rowCount()-1)
    {
        ui->participantsTable->model()->insertRow(row+1);

        QPushButton* delButton = new QPushButton();
        delButton->setText("Delete");

        ui->participantsTable->setIndexWidget(ui->participantsTable->model()->index(row , 2), delButton);

        connect(delButton, &QPushButton::released, [=]{
            handleDeleteButton(row);
        });
    }
}

void EditCompetitorsDialog::handleDeleteButton(int row)
{
    ui->participantsTable->model()->removeRow(row);
}

QDialogButtonBox* EditCompetitorsDialog::buttonBox()
{
    return ui->buttonBox;
}

QList<ParticipantInfo> EditCompetitorsDialog::participants()
{
    QList<ParticipantInfo> retval;
    for (int i = 0 ; i < ui->participantsTable->model()->rowCount()-1 ; i++)
    {
        QString name = ui->participantsTable->model()->data(ui->participantsTable->model()->index(i,0)).toString();
        QString carName = ui->participantsTable->model()->data(ui->participantsTable->model()->index(i,1)).toString();

        retval.append( ParticipantInfo(name, carName, i+1));
    }

    return retval;
}

File: Python/editcompetitorsdialog.py
from PyQt5 import QtWidgets, uic, QtCore, QtSerialPort
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtCore import *
from sys import platform

from resources import *
from pwdtimer_communications import *

class ParticipantInfo:
    participantName = ''
    carName = ''
    carNumber = ''
    raceTimes = []

    def __init__(self, participant_name='', car_name='', car_number=-1, race_times=[]) -> None:
        self.participantName = participant_name
        self.carName = car_name
        self.carNumber = car_number
        self.raceTimes = race_times
        

class EditCompetitorsDialog(QDialog):
    def __init__(self, competitor_list=[], parent=None):
        QDialog.__init__(self, parent)
        uic.loadUi(resource_path('editcompetitorsdialog.ui'),self)

        self.competitors = competitor_list
        self.updateCompetitorsTable()

    def updateCompetitorsTable(self):
        # Create the header row
        model = QStandardItemModel(len(self.competitors)+1,3,self)
        model.setHorizontalHeaderItem(0, QStandardItem("Name"))
        model.setHorizontalHeaderItem(1, QStandardItem("Car Name"))
        model.setHorizontalHeaderItem(2, QStandardItem("Delete"))

        # Fill in the existing data
        for K,competitor in enumerate(self.competitors):
            model.setData(model.index(K,0), competitor.participantName)
            model.setData(model.index(K,1), competitor.carName)        
        self.participantsTable.setModel(model)

        # Add the buttons for deleting a row
        for K,competitor in enumerate(self.competitors):
            part = self.competitors[K]

            delButton = QPushButton()
            delButton.setText("Delete")

            self.participantsTable.setIndexWidget(model.index(K,2), delButton)

            delButton.clicked.connect(lambda state, x=K: self.on_delete_button_pushed(x))

        self.participantsTable.model().dataChanged.connect(self.on_data_changed)
        # connect(ui->participantsTable->model(), SIGNAL(dataChanged(const QModelIndex&, const QModelIndex&)), this, SLOT(onDataChanged(const QModelIndex&, const QModelIndex&)));

    def on_delete_button_pushed(self,idx):
        self.participantsTable.model().removeRow(idx)

    def on_data_changed(self, tleft, bright):
        row = tleft.row()
        col = tleft.column()

        model = self.participantsTable.model()
        if row == model.rowCount()-1:
            model.insertRow(row+1)

            delButton = QPushButton()
            delButton.setText("Delete")
            
            self.participantsTable.setIndexWidget(model.index(row,2), delButton)
            delButton.clicked.connect(lambda state, x=row: self.on_delete_button_pushed(x))

    def participants(self):
        participants = []
        model = self.participantsTable.model()
        for I in range(model.rowCount()-1):
            name = model.data(model.index(I,0))
            car_name = model.data(model.index(I,1))
            participants.append( ParticipantInfo(name,car_name, I+1))
        return participants

